`ot_lc_ctrl`: Fix HW_CFG load memcpy size calculation

AlexJones0 opened the pull request:
These `memcpy`s were accidentally using the size of the dereferenced byte array (i.e., always just 1 byte), and so were only copying the first byte of the HW_CFG device ID and manufacturing state that were being loaded from the OTP.

AlexJones0 committed:
[ot] hw/opentitan: ot_lc_ctrl: Fix HW_CFG load memcpy size calculation

These `memcpy`s were accidentally using the size of the dereferenced
byte array (i.e., always just 1 byte), and so were only copying the
first byte of the HW_CFG device ID and manufacturing state that were
being loaded from the OTP.

Also add some local calculations and static assertions to check that the
HW_CFG value sizes that are given do not become out of sync with the
size of the registers.

Signed-off-by: Alex Jones <alex.jones@lowrisc.org>

rivos-eblot reviewed (approved):
Shouldn't we add static_assert to ensure that for some reason the size of those arrays change, they cannot overflow the dedicated registers?

AlexJones0 commented:
> Shouldn't we add static_assert to ensure that for some reason the size of those arrays change, they cannot overflow the dedicated registers?

Yes, I will add that to this PR.

*Edit*: Pushed that now. Not sure if there is a cleaner way to do this, feedback welcome :)

rivos-eblot commented on hw/opentitan/ot_lc_ctrl.c:
```diff
@@ -146,6 +146,15 @@ REG32(MANUF_STATE_7, 0x88u)
 #define LC_TOKEN_DWORDS         (LC_TOKEN_WIDTH / sizeof(uint64_t))
 
+#define LC_LAST_DEV_ID_REG   (R_DEVICE_ID_7)
+#define LC_DEV_ID_REGS_COUNT (LC_LAST_DEV_ID_REG - R_DEVICE_ID_0 + 1u)
```
Is REGS vs. REG on the next group of macros expected?

AlexJones0 replied:
No, this was unintentional. This is removed when using `OT_REG_SPAN` as you suggested below.

rivos-eblot commented:
> _Edit_: Pushed that now. Not sure if there is a cleaner way to do this, feedback welcome :)

As we use this idiom in several files, it might be worth to define something like:
````h
#define OT_REG_NAME_IDX(_n_, _i_) (R_ ## _n_ ## _ ## _i_)
#define OT_REG_COUNT(_n_, _l_)    (OT_REG_NAME_IDX(_n_, _l_) - OT_REG_NAME_IDX(_n_, 0) + 1u)
#define OT_REG_SPAN(_n_, _l_)     (OT_REG_COUNT(_n_, _l_) * sizeof(uint32_t))

// example
#define LC_MANUF_STATE_WIDTH OT_REG_SPAN(MANUF_STATE, 7)
#define LC_DEVICE_ID_WIDTH   OT_REG_SPAN(DEVICE_ID, 7)

````
in `ot_common.h`?

AlexJones0 committed:
[ot] hw/opentitan: ot_common: Add register span macros

Adds macros for generic computing of the size of register spans for
OpenTitan multi registers, which can be used to e.g. make static
assertions to ensure that the size of macros matches the size of
buffers being used.

Co-authored-by: Emmanuel Blot <eblot@rivosinc.com>
Signed-off-by: Alex Jones <alex.jones@lowrisc.org>
